googlefirestore/src/main/java/com/yahoo/ycsb/db/firestore/GoogleFirestoreClient.java

```diff
@@ -0,0 +1,246 @@
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+package com.yahoo.ycsb.db.firestore;
+
+import com.google.api.core.ApiFuture;
+import com.google.auth.oauth2.GoogleCredentials;
+import com.google.cloud.firestore.DocumentReference;
+import com.google.cloud.firestore.DocumentSnapshot;
+import com.google.cloud.firestore.Firestore;
+import com.google.cloud.firestore.FirestoreOptions;
+import com.google.cloud.firestore.Query;
+import com.google.cloud.firestore.QuerySnapshot;
+import com.google.cloud.firestore.SetOptions;
+import com.google.cloud.firestore.WriteResult;
+import com.yahoo.ycsb.ByteIterator;
+import com.yahoo.ycsb.DB;
+import com.yahoo.ycsb.DBException;
+import com.yahoo.ycsb.Status;
+import com.yahoo.ycsb.StringByteIterator;
+
+import org.apache.log4j.Level;
+import org.apache.log4j.Logger;
+
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Set;
+import java.util.Vector;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * YCSB Client for Google's Cloud Firestore.
+ */
+public class GoogleFirestoreClient extends DB {
+  /**
+   * The names of properties which can be specified in the config files and flags.
+   */
+  public static final class GoogleFirestoreProperties {
+    private GoogleFirestoreProperties() {}
+
+    /**
+     * The Cloud Firestore project ID to use when running the YCSB benchmark.
+     */
+    static final String PROJECT = "googlefirestore.projectId";
+    static final String PRIVKEYFILE = "googlefirestore.serviceAccountKey";
+    static final String DEBUG = "googlefirestore.debug";
+  }
+
+  private static final Logger LOGGER = Logger.getLogger(GoogleFirestoreClient.class);
+
+  private Firestore fsDb;
+
+  @Override
+  public void init() throws DBException {
+    if (fsDb != null) {
+      return;
+    }
+    Properties properties = getProperties();
+    String debug = properties.getProperty(GoogleFirestoreProperties.DEBUG);
+    if (null != debug && "true".equalsIgnoreCase(debug)) {
+      LOGGER.setLevel(Level.DEBUG);
+    }
+    String projectId = properties.getProperty(GoogleFirestoreProperties.PROJECT);
+    if (projectId == null) {
+      throw new DBException("Must provide project ID.");
+    }
+    String privateKeyFile = properties.getProperty(GoogleFirestoreProperties.PRIVKEYFILE);
+    if (privateKeyFile == null) {
+      throw new DBException("Must provide full path to private key file.");
+    }
+    try {
+      GoogleCredentials gCreds = GoogleCredentials.fromStream(new FileInputStream(privateKeyFile));
+      FirestoreOptions fsOptions = FirestoreOptions.newBuilder().setCredentials(gCreds).build();
+      fsDb = fsOptions.getService();
+    } catch (FileNotFoundException e) {
+      LOGGER.error("Can't find key.", e);
```

Comment:
In the case of one of these exceptions, shouldn't we throw a DBException to bring things to a stop?

Reply:
Fixed.